clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1256,65 +1182,331 @@ static void populateIntHeader(SYCLIntegrationHeader &H, const StringRef Name,
-}
+public:
+  // Mark these virutal so that we can use override in the implementer classes,
+  // despite virtual dispatch never being used.
+  virtual void handleSyclAccessorType(const FieldDecl *, QualType) {}
+  virtual void handleSyclSamplerType(const FieldDecl *, QualType) {}
+  virtual void handleSyclSpecConstantType(const FieldDecl *, QualType) {}
+  virtual void handleStructType(const FieldDecl *, QualType) {}
+  virtual void handleReferenceType(const FieldDecl *, QualType) {}
+  virtual void handlePointerType(const FieldDecl *, QualType) {}
+  virtual void handleArrayType(const FieldDecl *, QualType) {}
+  virtual void handleScalarType(const FieldDecl *, QualType) {}
+  // Most handlers shouldn't be handling this, just the field checker.
+  virtual void handleOtherType(const FieldDecl *, QualType) {}
+};
+
+// A type to check the valididty of all of the argument types.
+class SyclKernelFieldChecker
+    : public SyclKernelFieldHandler<SyclKernelFieldChecker> {
+  bool IsInvalid = false;
+  DiagnosticsEngine &Diag;
+
+public:
+  SyclKernelFieldChecker(Sema &S)
+      : SyclKernelFieldHandler(S), Diag(S.getASTContext().getDiagnostics()) {}
+  bool isValid() { return !IsInvalid; }
+
+  void handleReferenceType(const FieldDecl *FD, QualType ArgTy) final {
+    IsInvalid = Diag.Report(FD->getLocation(), diag::err_bad_kernel_param_type)
+                << ArgTy;
+  }
+  void handleStructType(const FieldDecl *FD, QualType ArgTy) final {
+    if (SemaRef.getASTContext().getLangOpts().SYCLStdLayoutKernelParams &&
+        !ArgTy->isStandardLayoutType())
+      IsInvalid =
+          Diag.Report(FD->getLocation(), diag::err_sycl_non_std_layout_type)
+          << ArgTy;
+    else {
+      CXXRecordDecl *RD = ArgTy->getAsCXXRecordDecl();
+      if (!RD->hasTrivialCopyConstructor())
+
+        IsInvalid =
+            Diag.Report(FD->getLocation(),
+                        diag::err_sycl_non_trivially_copy_ctor_dtor_type)
+            << 0 << ArgTy;
+      else if (!RD->hasTrivialDestructor())
+        IsInvalid =
+            Diag.Report(FD->getLocation(),
+                        diag::err_sycl_non_trivially_copy_ctor_dtor_type)
+            << 1 << ArgTy;
+    }
+  }
+
+  // We should be able to ahndle this, so we made it part of the visitor, but
+  // this is 'to be implemented'.
+  void handleArrayType(const FieldDecl *FD, QualType ArgTy) final {
+    IsInvalid = Diag.Report(FD->getLocation(), diag::err_bad_kernel_param_type)
+                << ArgTy;
+  }
+
+  void handleOtherType(const FieldDecl *FD, QualType ArgTy) final {
+    IsInvalid = Diag.Report(FD->getLocation(), diag::err_bad_kernel_param_type)
+                << ArgTy;
+  }
+};
+
+// A type that handles the accessor recursion and acts as a base for
+// SyclKernelDeclCreator. It doesn't 'own' anything other than the KernelObj
+// pointer and functionality required to add a param.
+class SyclKernelDeclBase : public SyclKernelFieldHandler<SyclKernelDeclBase> {
```

Comment:
I don't really get why do we need a base and a child class for kernel decl creation.
Could you please elaborate a bit more?